File: DSA/Linked_list_in_Python_and_C/Doubly_linked_list.c
"""
    Simple implementation of a doubly linked list in C
"""

#include <stdio.h>
#include <stdlib.h>

// Define a structure for a node in the doubly linked list
struct Node {
    int data;
    struct Node *prev;
    struct Node *next;
};

// Function to create a new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

int main() {
    // Creating nodes
    struct Node *head = createNode(1);
    struct Node *second = createNode(2);
    struct Node *third = createNode(3);

    // Connecting nodes forward
    head->next = second;
    second->next = third;
    third->next = NULL;

    // Connecting nodes backward
    third->prev = second;
    second->prev = head;
    head->prev = NULL;

    // Traversing and printing the linked list forward
    struct Node *current = head;
    printf("Forward Linked List: ");
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");

    // Traversing and printing the linked list backward
    current = third;
    printf("Backward Linked List: ");
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->prev;
    }
    printf("NULL\n");

    // Freeing memory
    free(head);
    free(second);
    free(third);

    return 0;
}

"""
Explanation:
- The `struct Node` represents a node in the doubly linked list. It contains an integer `data`, and two pointers `prev` and `next` to the previous and next nodes in the list respectively.
- The `createNode` function dynamically allocates memory for a new node, initializes its data, and sets its `prev` and `next` pointers to `NULL`.
- In the `main` function, three nodes are created with data values `1`, `2`, and `3`.
- The nodes are connected both forward and backward to form a doubly linked list by setting the `prev` and `next` pointers appropriately.
- The linked list is then traversed and printed both forward and backward.
- Finally, memory allocated for the nodes is freed using the `free` function.
"""

File: DSA/Linked_list_in_Python_and_C/Doubly_linked_list.py
"""Simple implementation of a doubly linked list in Python"""


class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
        new_node.prev = current

    def display_forward(self):
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

    def display_backward(self):
        current = self.head
        while current.next:
            current = current.next
        while current:
            print(current.data, end=" -> ")
            current = current.prev
        print("None")

# Creating a doubly linked list
dllist = DoublyLinkedList()
dllist.append(1)
dllist.append(2)
dllist.append(3)

# Displaying the doubly linked list forward and backward
print("Forward Linked List:")
dllist.display_forward()
print("Backward Linked List:")
dllist.display_backward()


"""
Explanation:
- The `Node` class represents a node in the doubly linked list. It contains a `data` attribute and `prev` and `next` attributes pointing to the previous and next nodes in the list, respectively.
- The `DoublyLinkedList` class represents the doubly linked list itself. It has a `head` attribute pointing to the first node in the list.
- The `append` method adds a new node with the given data to the end of the doubly linked list.
- The `display_forward` method traverses the doubly linked list forward and prints each node's data.
- The `display_backward` method traverses the doubly linked list backward and prints each node's data.
- In the `main` section, a doubly linked list object `dllist` is created, and nodes with data values `1`, `2`, and `3` are appended to it.
- Finally, the doubly linked list is displayed forward and backward using the `display_forward` and `display_backward` methods, respectively.
"""